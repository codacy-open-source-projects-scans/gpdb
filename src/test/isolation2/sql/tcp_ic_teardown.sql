-- Test ensuring that we perform a timed wait inside the TCP interconnect
-- teardown on the motion sender side, for the final response from the motion
-- receiver(s).
CREATE TABLE tcp_ic_teardown(i int);
INSERT INTO tcp_ic_teardown SELECT generate_series(1, 5);

SELECT gp_inject_fault('waitOnOutbound', 'suspend', dbid)
    FROM gp_segment_configuration WHERE role = 'p' AND content = 0;
SELECT gp_inject_fault('doSendStopMessageTCP', 'suspend', dbid)
    FROM gp_segment_configuration WHERE role = 'p' AND content = -1;

1: SET gp_interconnect_transmit_timeout TO '3s';
-- Use a LIMIT to squelch the motion node in order to send a 'stop' message.
1&: SELECT * FROM tcp_ic_teardown LIMIT 1;

-- Ensure that we have suspended the QD's gather motion receiver at the point
-- before it sends out the 'stop' message and have reached the point just prior
-- to starting the timed wait during TCP teardown on one of the motion senders.
SELECT gp_wait_until_triggered_fault('waitOnOutbound', 1, dbid)
    FROM gp_segment_configuration WHERE role = 'p' AND content = 0;
SELECT gp_wait_until_triggered_fault('doSendStopMessageTCP', 1, dbid)
    FROM gp_segment_configuration WHERE role = 'p' AND content = -1;

-- Let the timed wait proceed on the sender side.
SELECT gp_inject_fault('waitOnOutbound', 'reset', dbid)
    FROM gp_segment_configuration WHERE role = 'p' AND content = 0;

!\retcode sleep 6;

SELECT gp_inject_fault('doSendStopMessageTCP', 'reset', dbid)
    FROM gp_segment_configuration WHERE role = 'p' AND content = -1;
-- After 6s have elapsed (enough to have covered the timed wait of 3s, we should
-- have consequently ERRORed out on the motion sender side)
1<:
