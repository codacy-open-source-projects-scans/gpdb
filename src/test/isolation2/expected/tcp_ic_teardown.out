-- Test ensuring that we perform a timed wait inside the TCP interconnect
-- teardown on the motion sender side, for the final response from the motion
-- receiver(s).
CREATE TABLE tcp_ic_teardown(i int);
CREATE TABLE
INSERT INTO tcp_ic_teardown SELECT generate_series(1, 5);
INSERT 0 5

SELECT gp_inject_fault('waitOnOutbound', 'suspend', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content = 0;
 gp_inject_fault 
-----------------
 Success:        
(1 row)
SELECT gp_inject_fault('doSendStopMessageTCP', 'suspend', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content = -1;
 gp_inject_fault 
-----------------
 Success:        
(1 row)

1: SET gp_interconnect_transmit_timeout TO '3s';
SET
-- Use a LIMIT to squelch the motion node in order to send a 'stop' message.
1&: SELECT * FROM tcp_ic_teardown LIMIT 1;  <waiting ...>

-- Ensure that we have suspended the QD's gather motion receiver at the point
-- before it sends out the 'stop' message and have reached the point just prior
-- to starting the timed wait during TCP teardown on one of the motion senders.
SELECT gp_wait_until_triggered_fault('waitOnOutbound', 1, dbid) FROM gp_segment_configuration WHERE role = 'p' AND content = 0;
 gp_wait_until_triggered_fault 
-------------------------------
 Success:                      
(1 row)
SELECT gp_wait_until_triggered_fault('doSendStopMessageTCP', 1, dbid) FROM gp_segment_configuration WHERE role = 'p' AND content = -1;
 gp_wait_until_triggered_fault 
-------------------------------
 Success:                      
(1 row)

-- Let the timed wait proceed on the sender side.
SELECT gp_inject_fault('waitOnOutbound', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content = 0;
 gp_inject_fault 
-----------------
 Success:        
(1 row)

!\retcode sleep 6;
-- start_ignore

-- end_ignore
(exited with code 0)

SELECT gp_inject_fault('doSendStopMessageTCP', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content = -1;
 gp_inject_fault 
-----------------
 Success:        
(1 row)
-- After 6s have elapsed (enough to have covered the timed wait of 3s, we should
-- have consequently ERRORed out on the motion sender side)
1<:  <... completed>
ERROR:  timed out waiting for response from motion receiver during TCP interconnect teardown  (seg1 slice1 192.168.0.148:7003 pid=654372)
DETAIL:  1 connection(s) with pending response after 3 seconds
